pkg/querier/frontend/worker.go

```diff
@@ -118,116 +114,89 @@ func (w *worker) watchDNSLoop(servCtx context.Context) error {
 			switch update.Op {
 			case naming.Add:
 				level.Debug(w.log).Log("msg", "adding connection", "addr", update.Addr)
-				ctx, cancel := context.WithCancel(servCtx)
-				cancels[update.Addr] = cancel
-				w.runMany(ctx, update.Addr)
+				client, err := w.connect(servCtx, update.Addr)
+				if err != nil {
+					level.Error(w.log).Log("msg", "error connecting", "addr", update.Addr, "err", err)
+				}
+
+				w.managers[update.Addr] = newFrontendManager(servCtx, w.log, w.server, client, w.cfg.GRPCClientConfig)
 
 			case naming.Delete:
 				level.Debug(w.log).Log("msg", "removing connection", "addr", update.Addr)
-				if cancel, ok := cancels[update.Addr]; ok {
-					cancel()
+				if mgr, ok := w.managers[update.Addr]; ok {
+					mgr.stop()
```

Comment:
Since we only stop the manager here, it's still in the map. We use the length of the map to calculate the currency in the functions below this. Wouldn't we need to account for stopped managers?

Reply:
Dang, good catch.  Pushing a fix that removes the stopped manager from the map.